### PR TITLE
Add an option to merge intervals for better GGVCFs performance on GDB exome input

### PR DESCRIPTION
**Pull request by ldgauthier**
Rather than start off the exome genotyping task with a step to merge the query intervals for better GDB performance, make it an option as for Import in #5540 

This is necessary to get the exome joint calling pipeline onto official GATK release dockers.

### CODE REVIEW

**Comment by ldgauthier** [comment]
Tests pass!  (after rebasing to remove the offending ReadsPipeline Spark test)

**Comment by ldgauthier** [comment]
@droazen can you reassign this?  I'm guessing it's going to be a while before Louis gets to it and this is a feature we're using in production.  It would be great to get this in the release.

**Comment by droazen** [comment]
@jamesemery can you review this one when you get a chance? It's only ~100 lines or so.

File: src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java [hunk 1]
```diff
@@ -157,7 +157,7 @@
      * Walker base classes (ReadWalker, etc.) are responsible for hooking these intervals up to
      * their particular driving data source.
      */
-    List<SimpleInterval> userIntervals;
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java]: We would prefer not to expose these engine fields for safety purposes. As of right now these fields are only exposed to engine walkers etc... I would say revert this change.

File: src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java [hunk 2]
```diff
@@ -432,7 +432,7 @@ void initializeFeatures() {
      * Package-private so that engine classes can access it, but concrete tool child classes cannot.
      * May be overridden by traversals that require custom initialization of intervals.
      */
-    void initializeIntervals() {
+    protected void initializeIntervals() {
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java]: So we don't necessarily want to expose the `initializeIntervals()` argument to tool authors (mostly because we don't want to worry about tool authors changing the internal user intervals in unsafe ways. Since what you primarily need right here is to maybe mutate the user intervals on this line: `userIntervals = intervalArgumentCollection.getIntervals(sequenceDictionary);` I would consider adding a seperate method to do this.

Add a protected method `transformTraversalIntervals(getIntervals, sequenceDictionary)` that returns a list of intervals and make that what GATKTool sets as `userIntervals`. Then you can add the mutation you need to genomicsDBImport and leave the default behavior untouched without exposing anything dangerously by just making the default behavior return the supplied interval list.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java [hunk 3]
```diff
@@ -189,13 +217,22 @@ public void onTraversalStart() {
 
         final VCFHeader inputVCFHeader = getHeaderForVariants();
 
+        if (hasUserSuppliedIntervals()) {
+            if (mergeInputIntervals) {
+                intervals = intervalArgumentCollection.getSpanningIntervals(getBestAvailableSequenceDictionary());
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFs.java]: This should probably be shared code across the two places (here and `initializeIntervals()`). Alternatively you could add a safe getter to `userIntervals` that will let you access these intervals after the merge.

File: src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java (hunk 1 above)
Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java]: hmm... it looks like the tests/etc would be made somewhat easier if there were a getter for this. It should be harmless to add a `getUserIntervals()` that just returns an unmodifiable list of these simple intervals.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java [hunk 4]
```diff
@@ -232,6 +245,28 @@ public void assertMatchingGenotypesFromTileDB(File input, File expected, Locatab
...
+                .addArgument("V", genomicsDBUri)
+                .addOutput(output);
+        intervals.forEach(args::addInterval);
+        args.add("--" + GenomicsDBImport.MERGE_INPUT_INTERVALS_LONG_NAME);
+        args.add("--only-output-calls-starting-in-intervals");
+
+        Utils.resetRandomGenerator();
+        runCommandLine(args);
+
+        // if this isn't working it will take *FOREVER*
```
Comment (jamesemery) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java]: I don't like this as a solution for our tests. This could easily cause inexplicable failures on travis that nobody knows how to solve. Furthermore using the test timeout functionality on testNG seems like a crude solution. I think you should expose the userIntervalList with a getter and then use that to explicitly assert that the intervals were merged correctly.

Comment (ldgauthier) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java]: I do need to test that the mergeIntervals arg works as expected. I added an actual comparison.  If you're still concerned about the runtime if something breaks, then I can reduce the number of intervals.

File: src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java (hunk 2 above)
Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java]: Can you revert the change of `initializeIntervals()` to protected?

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java [hunk 5]
```diff
@@ -232,6 +245,32 @@ public void assertMatchingGenotypesFromTileDB(File input, File expected, Locatab
         runGenotypeGVCFSAndAssertSomething(genomicsDBUri, expected, NO_EXTRA_ARGS, VariantContextTestUtils::assertVariantContextsHaveSameGenotypes, reference);
     }
 
+    @Test(dataProvider = "getGVCFsForGenomicsDBOverMultipleIntervals")
+    public void testGenotypeGVCFsMultiIntervalGDBQuery(File input, File expected, List<Locatable> intervals, String reference) throws IOException {
```
Comment (jamesemery) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java]: If you really expect this test to be quick and only take 0.06 seconds we should probably have a contingency plan in place so this fails gracefully on travis (my fear is of a world where this fails and the signal is travis times out due to inactivity). You could add to this test a testNG timeout option ` timeOut = 1000` which would be 1000 ms or some more reasonable gross overestimation of the runtime that we expect to fail if this branch didin't work and use that.

File: src/test/java/org/broadinstitute/hellbender/engine/GATKToolUnitTest.java [hunk 6]
```diff
@@ -997,4 +1010,52 @@ public void testGetTraversalIntervals(@Nullable String intervals, Integer expect
...
+        return new Object[][]{
+                {Arrays.asList("1:21-21"), 1, Arrays.asList("1"), Arrays.asList(21), Arrays.asList(21), hg19MiniReference},
+                {Arrays.asList("1:21-21", "1:23-23", "1:200-300", "1:10000-16000"), 1, Arrays.asList("1"), Arrays.asList(21), Arrays.asList(16000), hg19MiniReference},
+                {Arrays.asList("1:21-21", "2:42-45", "2:10000-12345", "3:33-333"), 3, Arrays.asList("1","2","3"), Arrays.asList(21,42,33), Arrays.asList(21, 12345, 333), hg19MiniReference},
+                {null, 4, Arrays.asList("1","2","3","4"), Arrays.asList(1,1,1,1), Arrays.asList(16000,16000,16000,16000), hg19MiniReference},
+                {null, null, null, null, null, null}
+        };
+    }
+
+    @Test(dataProvider = "TestGetTransformedTraversalIntervalsProvider")
```
Comment (jamesemery) [inline src/test/java/org/broadinstitute/hellbender/engine/GATKToolUnitTest.java]: I like this test

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java [hunk 7]
```diff
@@ -238,7 +238,7 @@ public void testGenotypesOnly(File input, File expected, List<String> extraArgs,
     }
 
     //this only tests single-sample
-    @Test(dataProvider = "getGVCFsForGenomicsDB", timeOut = 1000)
+    @Test(dataProvider = "getGVCFsForGenomicsDB", timeOut = 1000000)
```
Comment (jamesemery) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java]: Whoops! Looks like I gave you a poor choice of number. I forgot to account for the runtime units being in terms of minutes...

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java (hunk 5 above)
Comment (jamesemery) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/GenotypeGVCFsIntegrationTest.java]: Whoops! Looks like I gave you a poor choice of number. I forgot to account for the runtime units being in terms of minutes. Sorry about that...
